utils/AREnvelope.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+#ifndef DSY_AR_ENVELOPE_H
+#define DSY_AR_ENVELOPE_H
+
+#include <stdlib.h>
+
+enum AREnvelopeState
+{
+    ARENV_ATTACK,
+    ARENV_ON,
+    ARENV_RELEASE,
+    ARENV_OFF
+};
+
+// A simple Attack-Release envelope.
+// Ramp a binary signal smoothly with attack and release times.
+// Must call Process for each sample with the current gate state.
+class AREnvelope
+{
+public:
+    AREnvelope() {}
+    ~AREnvelope() {}
+
+    void Init(int sampleRate, float attackTimeSeconds, float releaseTimeSeconds)
+    {
+        sampleRate_ = sampleRate;
+        state_ = ARENV_OFF;
+        sampleCounter_ = 0;
+        SetAttackTime(attackTimeSeconds);
+        SetReleaseTime(releaseTimeSeconds);
+    }
+
+    void SetAttackTime(float attackTimeSeconds)
+    {
+        attackSamples_ = attackTimeSeconds * sampleRate_;
+        if (sampleCounter_ > attackSamples_)
+            sampleCounter_ = attackSamples_; // deal with jumps
+    }
+
+    void SetReleaseTime(float releaseTimeSeconds)
+    {
+        releaseSamples_ = releaseTimeSeconds * sampleRate_;
+        if ((releaseSamples_ - sampleCounter_) < 0)
+            sampleCounter_ = 0; // deal with jumps
```

Comment:
Absolute nitpick, feel free to ignore: changing the slope this way introduces a discontinuity.

I think a better way to deal with jumps would be to measure how far into the release you are already and then pick up from that point with the new slope. This changes a discontinuity into a cusp.

```C
   float pctInto = ((float)sampleCounter_) / ((float)(releaseSamples_));
   releaseSamples_ = releaseTimeSeconds * sampleRate_;
   sampleCounter_ = pctInto * (float)releaseSamples_;
```

Without this code, there may be a teeny pop in the audio if you change the attack/release time during the ramp. Which seems like an edge case.

Because this just downgrades a 0th order discontinuity (discontinuity) to a first order discontinuity (cusp), there will still be audible artifacts, they'll just be quieter.